Source/loadsave.cpp

```diff
@@ -1677,20 +1681,43 @@ bool IsHeaderValid(uint32_t magicNumber)
 	return false;
 }
 
+// Returns the size of the hotkeys file with the number of hotkeys passed and if a header with the number of hotkeys is present in the file
+size_t HotkeysSize(size_t nHotkeys=NUM_HOTKEYS, bool header=1) {
+	return (sizeof(uint8_t)*header) + (nHotkeys * sizeof(int32_t)) + (nHotkeys * sizeof(int8_t) + sizeof(int32_t) + sizeof(int8_t));
+}
+
 void LoadHotkeys()
 {
 	LoadHelper file("hotkeys");
-	if (!file.IsValid())
-		return;
 
 	auto &myPlayer = Players[MyPlayerId];
+	size_t nHotkeys = 4;
 
-	for (auto &spellId : myPlayer._pSplHotKey) {
-		spellId = static_cast<spell_id>(file.NextLE<int32_t>());
+	// Checking if the save file has the old format with only 4 hotkeys
+	if (!file.IsValid(HotkeysSize(4,0))) {
+		// If it doesn't match exactly the size expected for 4 hotkeys, assume we have a new style format and read the number of hotkeys from the beginning of the file 
+		nHotkeys = file.NextLE<uint8_t>();
 	}

```

Comment:
FileHelper::IsValid() returns true if the file is >= the provided size. This check doesn't function as intended since it expects a strict equals comparison.

You'll need to structure it so the check is positive:
```suggestion
	if (file.IsValid(HotkeysSize(4, true))) {
		// The file contains a header byte and and at least 4 entries, so we can assume it's a new format save
		nHotkeys = file.NextLE<uint8_t>();
	}
```
This also removes the need to have the header flag parameter in `HotkeysSize()` since it's set to true in both call sites.

Eventually we could drop support for saves made in devilutionX before this PR is merged, so after the next release this could be inverted for an early return again.
```cpp
	if (!file.IsValid(HotkeysSize(4, true))) {
		// Missing header byte or empty hotkeys file, probably an old save.
		return;
	}

	nHotkeys = file.NextLE<uint8_t>();
```